src/rest_vol.c

```diff
@@ -732,6 +796,350 @@ H5_rest_set_connection_information(void)
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_PARSEERROR, FAIL, "JSON parse tree creation failed");
+
+            /* Retrieve the authentication message */
+            if (NULL == (key_obj = yajl_tree_get(parse_tree, ad_auth_message_keys, yajl_t_string)))
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_PARSEERROR, FAIL, "can't retrieve authentication instructions message");
+            if (NULL == (instruction_string = YAJL_GET_STRING(key_obj)))
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_PARSEERROR, FAIL, "can't retrieve authentication instructions message");
+
+            printf("Please follow the instructions below\n %s\n\nHit Enter when completed", instruction_string);
+            getchar();
+
+            /* Assume that the user has now properly signed in - attempt to retrieve token */
+
+            if (NULL == (key_obj = yajl_tree_get(parse_tree, device_code_keys, yajl_t_string)))
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_PARSEERROR, FAIL, "can't retrieve authentication device code");
+
+            if (NULL == (device_code = YAJL_GET_STRING(key_obj)))
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_PARSEERROR, FAIL, "can't retrieve authentication device code");
+#ifdef RV_CONNECTOR_DEBUG
+            printf("-> Device code: \"%s\"\n", device_code);
+#endif
+
+            /* Form URL from tenant ID string */
+            if (snprintf(tenant_string, sizeof(tenant_string),
+                    "https://login.microsoftonline.com/%s/oauth2/v2.0/token", ad_info->tenantID) < 0)
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_SYSERRSTR, FAIL, "snprintf error");
+
+            if (CURLE_OK != curl_easy_setopt(curl, CURLOPT_URL, tenant_string))
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_CANTSET, FAIL, "can't set cURL request URL: %s", curl_err_buf);
+
+            /* Form token string */
+            if (snprintf(data_string, sizeof(data_string),
+                    "grant_type=urn%%3Aietf%%3Aparams%%3Aoauth%%3Agrant-type%%3Adevice_code&device_code=%s&client_id=%s",
+                    device_code, ad_info->clientID) < 0)
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_SYSERRSTR, FAIL, "snprintf error");
+
+            if (CURLE_OK != curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data_string))
+                FUNC_GOTO_ERROR(H5E_VOL, H5E_CANTSET, FAIL, "can't set cURL POST data: %s", curl_err_buf);
+
+#ifdef RV_CONNECTOR_DEBUG
+            printf("-> Authentication server URL: \"%s\"\n", tenant_string);
+            printf("-> Request payload: \"%s\"\n", data_string);
+#endif
+            /* Request access token from the server */
+            CURL_PERFORM(curl, H5E_VOL, H5E_CANTGET, FAIL);
+
+            yajl_tree_free(parse_tree);
+            parse_tree = NULL;
+
+#ifdef RV_CONNECTOR_DEBUG
+            printf("-> Authentication successfully completed.\n");
+            printf("-> Authentication server response:\n-> %s\n", response_buffer.buffer);
+#endif
+
```

Comment:
It seems like everything below here is probably applicable to the unattended authentication case as well isn't it? Should it be moved outside of this block?

Reply:
Moved the first line out so it applies to both cases.

Reply:
Ah, maybe I'm wrong, but I meant that it seems like all the code for checking the validity of the access token and also retrieving the refresh token would also apply to the unattended case, but currently it looks like this only applies for the attended case's code. I'm guessing you'll probably have to move some stuff around when doing the token refreshing work anyway though, so this will work for now.